src/controller/SetUpCodePairer.cpp

```diff
@@ -40,61 +41,44 @@ using namespace chip::Tracing;
-    if (isQRCode)
-    {
-        ReturnErrorOnFailure(QRCodeSetupPayloadParser(setUpCode).populatePayload(payload));
-        VerifyOrReturnError(payload.isValidQRCodePayload(), CHIP_ERROR_INVALID_ARGUMENT);
-    }
-    else
-    {
-        ReturnErrorOnFailure(ManualSetupPayloadParser(setUpCode).populatePayload(payload));
-        VerifyOrReturnError(payload.isValidManualCode(), CHIP_ERROR_INVALID_ARGUMENT);
-    }
-
-    return CHIP_NO_ERROR;
-}
-} // namespace
-
 CHIP_ERROR SetUpCodePairer::PairDevice(NodeId remoteId, const char * setUpCode, SetupCodePairerBehaviour commission,
                                        DiscoveryType discoveryType, Optional<Dnssd::CommonResolutionData> resolutionData)
 {
     VerifyOrReturnErrorWithMetric(kMetricSetupCodePairerPairDevice, mSystemLayer != nullptr, CHIP_ERROR_INCORRECT_STATE);
     VerifyOrReturnErrorWithMetric(kMetricSetupCodePairerPairDevice, remoteId != kUndefinedNodeId, CHIP_ERROR_INVALID_ARGUMENT);
 
-    SetupPayload payload;
-    ReturnErrorOnFailure(GetPayload(setUpCode, payload));
+    std::vector<SetupPayload> payloads;
+    ReturnErrorOnFailure(SetupPayload::FromStringRepresentation(setUpCode, payloads));
 
-    if (resolutionData.HasValue())
+    if (resolutionData.HasValue() && payloads.size() == 1 && mSetupPayloads.size() == 1)
     {
         VerifyOrReturnErrorWithMetric(kMetricSetupCodePairerPairDevice, discoveryType != DiscoveryType::kAll,
                                       CHIP_ERROR_INVALID_ARGUMENT);
-        if (mRemoteId == remoteId && mSetUpPINCode == payload.setUpPINCode && mConnectionType == commission &&
+        if (mRemoteId == remoteId && mSetupPayloads[0].setUpPINCode == payloads[0].setUpPINCode && mConnectionType == commission &&
             mDiscoveryType == discoveryType)
         {
-            NotifyCommissionableDeviceDiscovered(resolutionData.Value());
+            // Not passing a discriminator is ok, since we have only one payload.
+            NotifyCommissionableDeviceDiscovered(resolutionData.Value(), std::nullopt);
             return CHIP_NO_ERROR;
         }
     }
 
+    ResetDiscoveryState();
+
     mConnectionType = commission;
     mDiscoveryType  = discoveryType;
     mRemoteId       = remoteId;
-    mSetUpPINCode   = payload.setUpPINCode;
-
-    ResetDiscoveryState();
+    mSetupPayloads  = std::move(payloads);
 
-    if (resolutionData.HasValue())
+    if (resolutionData.HasValue() && mSetupPayloads.size() == 1)
     {
-        NotifyCommissionableDeviceDiscovered(resolutionData.Value());
+        // No need to pass in a discriminator if we have only one payload, which
+        // is good because we don't have a full discriminator here anyway.
+        NotifyCommissionableDeviceDiscovered(resolutionData.Value(), std::nullopt);
         return CHIP_NO_ERROR;
     }
 
-    ReturnErrorOnFailureWithMetric(kMetricSetupCodePairerPairDevice, Connect(payload));
+    ReturnErrorOnFailureWithMetric(kMetricSetupCodePairerPairDevice, Connect());
```

Comment:
should we log something here? I undrstand here that we have a resolution data, but our mSetupPayloads is larger than 1 ... so we are saying we failed to pair?

Old code seems to say: if you have resolution data then we succeed, otherwise we fail. In the new code we say "if you have resolution data but more than one payload, we still fail as if we had no resolution data". Following what the code is doing through some logs seems beneficial.

Reply:
I don't follow the question.  If we have multiple payloads, then the fact that we were given only one resolution data is not enough to stop discovery, because we have multiple payloads.  This is not "failing", this is calling "do discovery", no?

Old code says: If you have resolution data, just use that, otherwise do discovery.

New code says: If you have resolution data (for one thing, by definition) and we were only looking for one thing, use that.  Otherwise, do discovery.

Reply:
I misread the code path here. This whole code does not feel very easy to follow/read :(